corehq/util/couch_doc_processor.py

```diff
@@ -0,0 +1,372 @@
+    about an iteration that is in progress. The state will be maintained
+    indefinitely unless it is removed with `discard_state()`.
+    :param chunk_size: Number of documents to yield before updating the
+    iteration checkpoint. In the worst case about this many documents
+    that were previously yielded may be yielded again if the iteration
+    is stopped and later resumed.
+    :param maintain_retry_state: True if the retry state of documents should be maintained
+    after completion of the iterator. If set to True then future runs of the same iterator
+    will continue to iterate over the retry queue until all documents
+    have been successfully processed.
+    """
+
+    def __init__(self, db, doc_types, iteration_key, chunk_size=100, maintain_retry_state=True):
+        if isinstance(doc_types, str):
+            raise TypeError("expected list of strings, got %r" % (doc_types,))
+        self.db = db
+        self.original_doc_types = doc_types = sorted(doc_types)
+        self.iteration_key = iteration_key
+        self.chunk_size = chunk_size
+        iteration_name = "{}/{}".format(iteration_key, " ".join(doc_types))
+        self.iteration_id = hashlib.sha1(iteration_name).hexdigest()
+        self.maintain_retry_state = maintain_retry_state
+        try:
+            self.state = db.get(self.iteration_id)
+        except ResourceNotFound:
+            # new iteration
+            self.state = {
+                "_id": self.iteration_id,
+                "doc_type": "ResumableDocsByTypeIteratorState",
+                "retry": {},
+                "retry_cleared": [],
+
+                # for humans
+                "name": iteration_name,
+                "timestamp": datetime.utcnow().isoformat()
+            }
+            args = {}
+        else:
+            # resume iteration
+            args = self.state.get("offset", {}).copy()
+            if args:
+                assert args.get("startkey"), args
+                doc_type = args["startkey"][0]
+                # skip doc types before offset
+                doc_types = doc_types[doc_types.index(doc_type):]
+            else:
+                # non-retry phase of iteration is complete
+                doc_types = []
+        args.update(
+            view_name='all_docs/by_doc_type',
+            chunk_size=chunk_size,
+            log_handler=ResumableDocsByTypeLogHandler(self),
+            include_docs=True,
+            reduce=False,
+        )
+        self.view_args = args
+        self.doc_types = doc_types
+
+    def __iter__(self):
+        args = self.view_args
+        for doc_type in self.doc_types:
+            if args.get("startkey", [None])[0] != doc_type:
+                args.pop("startkey_docid", None)
+                args["startkey"] = [doc_type]
+            args["endkey"] = [doc_type, {}]
+            for result in paginate_view(self.db, **args):
+                yield result['doc']
+
+        retried = {}
```

Comment:
This still relies on raising an exception to stop the iteration which is what I was trying to avoid but as you pointed out that breaks the iterator contract (which I'm not really concerned about). What this does add is a way to clean up any docs that have been retired (successfully or up to the max_retry limit). I'm not going to include this for now and have reverted my other changes. 

What I have done is give the doc processor class a chance to handle skipped docs: fbd8f83